2020/day-4-passport-processing/index.ts

```diff
@@ -0,0 +1,75 @@
+// byr (Birth Year) - four digits; at least 1920 and at most 2002.
+// iyr (Issue Year) - four digits; at least 2010 and at most 2020.
+// eyr (Expiration Year) - four digits; at least 2020 and at most 2030.
+// hgt (Height) - a number followed by either cm or in:
+//     If cm, the number must be at least 150 and at most 193.
+//     If in, the number must be at least 59 and at most 76.
+// hcl (Hair Color) - a # followed by exactly six characters 0-9 or a-f.
+// ecl (Eye Color) - exactly one of: amb blu brn gry grn hzl oth.
+// pid (Passport ID) - a nine-digit number, including leading zeroes.
+// cid (Country ID) - ignored, missing or not.
+
+const parseList = (list: string) => {
+  return list
+    .trim()
+    .split(/\n\n/)
+    .map(entry => entry.split(/\s+/).map(x => x.split(':')))
+    .map(Object.fromEntries)
+}
+
+type Passport = {
+  byr?: string
+  iyr?: string
+  eyr?: string
+  hgt?: string
+  hcl?: string
+  ecl?: string
+  pid?: string
+  cid?: string
+}
+
+type Maybe<T> = T | null | undefined
+
+const isWithinRange = (value: string | number, start: string | number, end: string | number) =>
+  Number(value) >= start && Number(value) <= end
+
+const isValidBirthYear = (byr: Maybe<string | number>) => !!byr && isWithinRange(byr, 1920, 2002)
+
+const isValidIssueYear = (iyr: Maybe<string | number>) => !!iyr && isWithinRange(iyr, 2010, 2020)
+
+const isValidExpirationYear = (eyr: Maybe<string | number>) => !!eyr && isWithinRange(eyr, 2020, 2030)
+
+const isValidHeight = (hgt: Maybe<string>) => {
+  const [_, value, unit] = hgt?.match(/([0-9]+)(cm|in)$/) || []
```

Comment:
What will `value` + `unit` be here if match fails? Also what is this actually destructing? Is it the matches? 

Final question, what is the `?` doing here?

Reply:
`value` and `unit` will both be undefined if there is no match:

<img width="479" alt="image" src="https://user-images.githubusercontent.com/24367010/103214126-8d59bf80-4907-11eb-8828-ccc9053d4736.png">

It's destructuring each match defined in the capture groups:

<img width="468" alt="image" src="https://user-images.githubusercontent.com/24367010/103214104-7d41e000-4907-11eb-9355-8ffd475121a2.png">

`.match` returns either `null` if there is no match, or an array of matches if successful. You'll get an error if you try to destructure `null`, which is why we need the ` || []` here in case of no match so we just get `undefined` properties instead of it blowing up:

<img width="441" alt="image" src="https://user-images.githubusercontent.com/24367010/103214170-a8c4ca80-4907-11eb-9a09-8df170633c03.png">

The `?` is a new JS feature called [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining), which allows you to access object properties that may not exist without having to do a bunch of checks:

<img width="538" alt="image" src="https://user-images.githubusercontent.com/24367010/103214755-3bb23480-4909-11eb-990b-1aa1d3af0b97.png">

It just returns `undefined` if the property doesn't exist, instead of blowing up